torchtune/modules/tokenizer.py

```diff
@@ -60,24 +67,46 @@ def encode(
         text: str,
         add_bos: bool = True,
         add_eos: bool = True,
+        trim_leading_whitespace: bool = False,
+        prefix: Optional[str] = None,
     ) -> List[int]:
         """Encode text into token IDs.
 
         Args:
             text (str): The input text to be encoded, unbatched.
             add_bos (bool): Whether to prepend BOS to the input, defaults to True.
             add_eos (bool): Whether to append EOS to the input, defaults to True.
-
+            trim_leading_whitespace (bool): Whether to trim leading whitespace from
+                underlying sentencepiece tokenization. Sentencepiece normally prepends
+                whitespace to any tokenized text, which can cause differences where
+                encode(s1) + encode(s2) != encode(s1 + s2) due to leading whitespace
+                added to s2. Default: False
```

Comment:
So when would I set this to True?

Reply:
I tried to describe in the comment on L138-L141. First, we only run into mismatches due to leading whitespace on the nth split of a string, where n>1. This is because sentencepiece automatically prepends whitespace to a string prior to encoding it, so when we do encode(s1+s2) and encode(s1) + encode(s2), s1 has been prepended with whitespace in both cases, and so the behavior is the same (this is point (a)).

Second, we also only see a mismatch if the tokenizer actually explicitly encodes whitespace, which not all tokenizers do. E.g. if `spm_test` corresponds to our tokenizer in `tests/assets/m.model`, and `spm_llama` corresponds to the usual Llama2 tokenizer, we can actually see that

```
>>> spm_test.encode(" ", add_bos=False, add_eos=False)
[]
>>> spm_test.encode("\n", add_bos=False, add_eos=False)
[]
>>> spm_llama.encode(" ", add_bos=False, add_eos=False)
[259]
>>> spm_llama.encode("\n", add_bos=False, add_eos=False)
[29871, 13],
```

so our test tokenizer does not return any tokens when it only sees whitespace. Because of this, we get different behavior in both cases when splitting a single string.
```
# When the tokenizer doesn't tokenize whitespace, the results match
>>> spm_test.encode("hi\nthere", add_bos=False, add_eos=False)
[476, 70]
>>> spm_test.encode("hi\n", add_bos=False, add_eos=False) + spm_test.encode("there", add_bos=False, add_eos=False)
[476, 70]

# On the regular Llama2 tokenizer, the results do not match
>>> spm_llama.encode("hi\nthere", add_bos=False, add_eos=False)
[7251, 13, 12711]
>>> spm_llama.encode("hi\n", add_bos=False, add_eos=False) + spm_llama.encode("there", add_bos=False, add_eos=False)
[7251, 13, 727]
```

Finally, there are some prompts that end with " " (e.g. our `GrammarErrorCorrectionTemplate`). In this case, the concatenated string actually contains a space, so we do not want to strip it from s2. But we do want to remove it from the preceding string to ensure we don't double-count it. This is the reason for the `.rstrip(" ")` in L150.

So these are the three conditions we need to check to determine whether to do this special handling, and they are all checked in `tokenize_messages` below.